OAuthSwift/OAuthSwiftHTTPRequest.swift

```diff
@@ -159,98 +123,150 @@ public class OAuthSwiftHTTPRequest: NSObject, NSURLSessionDelegate {
     }
 
 	public func cancel() {
-		task?.cancel()
+        // perform lock here to prevent cancel calls on another thread while creating the request
+        objc_sync_enter(self)
+        defer { objc_sync_exit(self) }
+
+        // either cancel the request if it's already running or set the flag to prohibit creation of the request
+        if let task = task {
+            task.cancel()
+        } else {
+            cancelRequested = true
+        }
 	}
 
-    public func makeRequest() throws -> NSMutableURLRequest {
-        return try OAuthSwiftHTTPRequest.makeRequest(self.URL, method: self.HTTPMethod, headers: self.headers, parameters: self.parameters, dataEncoding: self.dataEncoding, body: self.HTTPBody, paramsLocation: self.paramsLocation)
-    }
+}
 
-    public class func makeRequest(
-        URL: NSURL,
-        method: Method,
-        headers: [String : String],
-        parameters: Dictionary<String, AnyObject>,
-        dataEncoding: NSStringEncoding,
-        body: NSData? = nil,
-        paramsLocation: ParamsLocation = .AuthorizationHeader) throws -> NSMutableURLRequest {
-            let request = NSMutableURLRequest(URL: URL)
-            request.HTTPMethod = method.rawValue
-            return try setupRequestForOAuth(request,
-                headers: headers,
-                parameters: parameters,
-                dataEncoding: dataEncoding,
-                body: body,
-                paramsLocation: paramsLocation
-            )
-            
+// MARK: request preparation
+
+extension OAuthSwiftHTTPRequest {
+
+    private class func makeRequest(requestConfig: OAuthSwiftHTTPRequestConfig, credentials: OAuthSwiftCredential) throws -> NSMutableURLRequest {
+        let request = requestConfig.request.mutableCopy() as! NSMutableURLRequest
+        return try setupRequestForOAuth(request,
+            additionalParameters: requestConfig.additionalParameters,
+            dataEncoding: requestConfig.dataEncoding,
+            paramsLocation: requestConfig.paramsLocation,
+            credentials: credentials
+        )
     }
 
-    public class func setupRequestForOAuth(request: NSMutableURLRequest,
-        headers: [String : String],
-        parameters: Dictionary<String, AnyObject>,
+    private class func setupRequestForOAuth(request: NSMutableURLRequest,
+        additionalParameters: Dictionary<String, AnyObject>,
         dataEncoding: NSStringEncoding,
-        body: NSData? = nil,
-        paramsLocation : ParamsLocation = .AuthorizationHeader) throws -> NSMutableURLRequest {
-            for (key, value) in headers {
-                request.setValue(value, forHTTPHeaderField: key)
+        paramsLocation : ParamsLocation = .AuthorizationHeader,
+        credentials: OAuthSwiftCredential) throws -> NSMutableURLRequest {
+
+            // TODO: I think there might be a bug in the handling of the additionalParameters!
```

Comment:
Difficult to merge a big PR with many topic (see https://github.com/OAuthSwift/OAuthSwift/blob/master/CONTRIBUTING.md )
There is no issue referenced with some information about this bug